### ISSUE TITLE
Annotated Alias and Interaction with Special Forms

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3122]
**Describe the bug**

```python
from typing import TypeVar

from typing_extensions import Annotated, ClassVar

T = TypeVar('T')
Volatile = Annotated[T, None]

class Foo:
    x: Volatile[ClassVar[int]]

```

This produces an error of "ClassVar" is not allowed in this context. Annotated aliases should be able to wrap special forms.

Also in case relevant most of my real usages of this I import Volatile from a different module (not typing/typing_extensions). So typical usage is like,

runtime_aliases.py

```python
from typing_extensions import Annotated
from typing import TypeVar

T = TypeVar('T')
Volatile = Annotated[T, None]
```

toy.py

```python
from typing_extensions import ClassVar
from runtime_aliases import Volatile


class Foo:
    x: Volatile[ClassVar[int]]
```

**VS Code extension or command-line**
pyright 1.1.225

**Additional context**
I'm aware this is probably yet another edge case of all Annotated/Special Forms rules. While I use this a fair amount as second argument of annotated I may reuse metadata I need in several annotations, I do have work arounds I can use for ClassVar/Final by reversing order of annotated alias and classvar and then undoing reversal at runtime.

@GBeauregard does this work at runtime on master? I haven't checked if master runtime handles this case and whether it should.

**Comment by erictraut** [comment]
I think pyright is correct in this case. `Annotated[T, None]` is the equivalent of `T` from the perspective of a type checker, and `T[ClassVar[int]]` is nonsensical.

**Comment by hmc-cs-mdrissi** [comment]
That explanation is inconsistent with behavior pyright displays in this example. 

```python
from typing import TypeVar, List
from typing_extensions import Annotated

T = TypeVar('T')
Volatile = Annotated[T, None]

class Foo:
    x: Volatile[List[int]] # Both runtime and pyright are happy.
```

This code passes pyright even though if that argument applied it should fail in same way. So pyright is not treating Volatile as equivalent of just T. A pure typevar like here does fail,

```python
from typing import TypeVar, List
from typing_extensions import Annotated

T = TypeVar('T')
Volatile = T

class Foo:
    x: Volatile[List[int]] # error at runtime and pyright
```


Also runtime behavior when we ignore classvarness substitutes the type into annotated. So expected equivalence is Volatile[SomeType] -> Annotated[SomeType, None] -> SomeType. The first equivalence being runtime treatment, the second one being for type checker.

**Comment by erictraut** [comment]
`Annotated` is allowed to wrap `ClassVar`, but you cannot use `ClassVar` within other annotations or in the specialization of a generic type alias as you are doing here.

**Comment by hmc-cs-mdrissi** [comment]
Maybe this would be good to ask in typing-sig/typing about? I think reasons for Annotated being allowed to wrap ClassVar apply for aliases that simplify to Annotated[ClassVar[...]].

I can make thread tomorrow asking about this and what desired behavior should be.

**Comment by erictraut** [comment]
I don't plan to support this in pyright regardless, so I'd appreciate it if you wouldn't pursue this further. I've already gone way out of my way to accommodate `Annotated` as a wrapper in ways that I consider extreme edge cases for a very small number of users. There's a very simple workaround here: don't use `Annotated` in type aliases.

**Comment by hmc-cs-mdrissi** [comment]
Ok I will drop it. I'll note most of original argument for Annotated was actually for aliases as I use those more frequently then I do Annotated. But workarounds exist and I'm fine with them.
